modules/rte_mis_core/src/Plugin/Block/RoleBasedDetailsBlock.php

```diff
@@ -0,0 +1,599 @@
+        ->condition('status', 1)
+        ->accessCheck(FALSE)
+        ->condition('roles', 'block_admin');
+      return $query->count()->execute();
+    }
+
+    if ($current_role === 'district_admin') {
+      $currentUserId = $this->currentUser->id();
+      $currentUser = $this->entityTypeManager->getStorage('user')->load($currentUserId);
+
+      // Get location ID from user field.
+      $locationId = (int) $currentUser->get('field_location_details')->getString();
+
+      if (empty($locationId)) {
+        // No location ID found, return 0.
+        return 0;
+      }
+
+      // Load children taxonomy terms and count them.
+      $locationTree = $this->entityTypeManager->getStorage('taxonomy_term')->loadChildren($locationId);
+      return count($locationTree);
+    }
+
+    return 0;
+  }
+
+  /**
+   * Gets the count of schools based on the user's role and location.
+   *
+   * @param string $current_role
+   *   The role of the user ('state_admin', 'district_admin', 'block_admin').
+   * @param int|null $locationId
+   *   (Optional) The location ID to filter schools. If not provided, uses
+   *    current user's location.
+   *
+   * @return int
+   *   The count of schools based on the user's role and location.
+   */
+  public function getSchoolCount($current_role, ?int $locationId = NULL): int {
+    $matchingSchools = [];
+
+    // If $locationId is not provided, use the current user's location ID.
+    if ($locationId === NULL) {
+      $currentUserId = $this->currentUser->id();
+      $currentUser = $this->entityTypeManager->getStorage('user')->load($currentUserId);
+      $locationId = (int) $currentUser->get('field_location_details')->getString();
+    }
+
+    // Query all taxonomy terms in the 'school' vocabulary.
+    $term_storage = $this->entityTypeManager->getStorage('taxonomy_term');
+    $query = $term_storage->getQuery()
+      ->condition('vid', 'school')
+      ->accessCheck(FALSE);
+
+    $tids = $query->execute();
+
+    if (!empty($tids)) {
+      // Load all school terms.
+      $schools = $term_storage->loadMultiple($tids);
+
+      foreach ($schools as $school) {
+        // Get the field_location value.
+        $field_location = $school->get('field_location')->target_id;
+
+        // Load all parent terms of the school's location.
+        $locationTree = $term_storage->loadAllParents($field_location);
+        $locationIds = array_keys($locationTree);
+
+        // Check if the school's location matches the provided or current user's
+        // location.
+        if (in_array($locationId, $locationIds)) {
+          // Check user role to determine if we count this school.
+          if ($current_role == 'state_admin' ||
+                ($current_role == 'district_admin' && in_array($locationId, $locationIds)) ||
+                ($current_role == 'block_admin' && in_array($locationId, $locationIds))) {
+            $matchingSchools[] = $school;
+          }
+        }
+      }
+    }
```

Comment:
- State admin does have location defined in there profile.
- Can we simplify these condition as location check is common for district and block admin.